src/api/block_builder/input_selection/utxo_chains.rs

```diff
@@ -0,0 +1,215 @@
+                            }
+                        }
+                    }
+                }
+                Output::Nft(nft_output) => {
+                    // If the id is null then this output creates it and we can't have a previous output
+                    if !nft_output.nft_id().is_null() {
+                        let output_id = client.nft_output_id(*nft_output.nft_id()).await?;
+                        let output_response = client.get_output(&output_id).await?;
+                        if let OutputDto::Nft(nft_output) = &output_response.output {
+                            let nft_output = NftOutput::try_from(nft_output)?;
+
+                            let unlock_address = nft_output
+                                .unlock_conditions()
+                                .locked_address(nft_output.address(), current_time);
+
+                            utxo_chains.push((*unlock_address, output_response));
+                        }
+                    }
+                }
+                Output::Foundry(foundry_output) => {
+                    // if it's the first foundry output, then we can't have it as input
+                    if let Ok(output_id) = client.foundry_output_id(foundry_output.id()).await {
+                        let output_response = client.get_output(&output_id).await?;
+                        if let OutputDto::Foundry(foundry_output_dto) = &output_response.output {
+                            let foundry_output = FoundryOutput::try_from(foundry_output_dto)?;
+                            utxo_chains.push((Address::Alias(*foundry_output.alias_address()), output_response));
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        // Get recursively owned alias or nft outputs
+        get_alias_and_nft_outputs_recursively(self.client, &mut utxo_chains).await?;
+
+        let mut utxo_chain_inputs = Vec::new();
+        for (unlock_address, output_response) in utxo_chains {
+            let (address_index, internal) = match self.secret_manager {
+                Some(secret_manager) => {
+                    match unlock_address {
+                        Address::Ed25519(_) => {
+                            search_address(
+                                secret_manager,
+                                &bech32_hrp,
+                                self.coin_type,
+                                self.account_index,
+                                self.input_range.clone(),
+                                &unlock_address,
+                            )
+                            .await?
+                        }
+                        // Alias and NFT addresses can't be generated from a private key
+                        _ => (0, false),
+                    }
+                }
+                None => (0, false),
+            };
+
```

Comment:
Not sure I fully understand, why are we using 0 then ?

Reply:
Hm not sure anymore, copied from somewhere else
Should probably be changed to something which results in `chain: None` below